File: winning_moves.py
from typing import List, Tuple, Dict
from chessball_board import ChessBallBoard, Player, PieceType, possible_moves

def winning_moves(position: ChessBallBoard, player: Player) -> List[Dict]:
    """
    Returns the list of moves for `player` in `position` such that after the move,
    the ball is in the winning row (row 0 for black, row 5 for white).
    Each move is represented as a dict (as in possible_moves).
    """
    winner_row = 0 if player == Player.BLACK else position.ROWS - 1

    moves_and_boards = possible_moves(position, player)
    win_moves = []
    for move, board in moves_and_boards:
        ball_pos = board.find_ball()
        if ball_pos and ball_pos[0] == winner_row:
            win_moves.append(move)
    return win_moves

File: rust_chessball/src/winning_moves.rs
//! Winning move detection.

use crate::board::ChessBallBoard;
use crate::board::Player;
use crate::moves::{MoveInfo, possible_moves};

/// Return the list of moves for `player` that result in the ball being in the player's goal row.
///
/// For Black the winning row is 0; for White it's rows-1.
pub fn winning_moves(position: &ChessBallBoard, player: Player) -> Vec<MoveInfo> {
    let winner_row = match player {
        Player::Black => 0usize,
        Player::White => position.rows - 1,
        Player::Neutral => position.rows - 1, // neutral not used
    };
    let mut wins = Vec::new();
    for (mv, board_after) in possible_moves(position, player) {
        if let Some(ball_coord) = board_after.find_ball()
            && ball_coord.r == winner_row as isize
        {
            wins.push(mv);
        }
    }
    wins
}
